ambersim/rl/networks.py

```diff
@@ -0,0 +1,117 @@
+from typing import Sequence
+
+import flax.linen as nn
+import jax
+import jax.numpy as jnp
+from brax.training import distribution, networks, types
+from brax.training.agents.ppo.networks import PPONetworks
+from flax import struct
+
+"""
+Convienience tools for defining neural networks for RL.
+"""
+
+
+class MLP(nn.Module):
+    """A simple pickle-able multi-layer perceptron.
+
+    Args:
+        layer_sizes: Sizes of all hidden layers and the output layer.
+        activate_final: Whether to apply an activation function to the output.
+        bias: Whether to use a bias in the linear layers.
+    """
+
+    layer_sizes: Sequence[int]
+    activate_final: bool = False
+    bias: bool = True
+
+    @nn.compact
+    def __call__(self, x):
+        """Forward pass through the network."""
+        # TODO(vincekurtz): consider using jax control flow here
+        for i, layer_size in enumerate(self.layer_sizes):
```

Comment:
Agree on the TODO: unrolled for loops compile super slowly, so this is definitely going to incur some tech debt in a stack that already needs a while to compile.

I would recommend just running copilot on it (there should be some vim extensions that allow it), since it's pretty good at converting control flow stuff into jax-compliant code. I can also pick it up if you want.

Reply:
Alternatively, give it to chatgpt

Reply:
So I gave it a (brief) shot with copilot, and was unable to get things to works. Seems like there are some nontrivial interactions between flax's `nn.compact` flag and jax's `fori_loop` and `scan`. Possibly the same reason that the brax MLP uses something similar?

Reply:
hmm ok i didn't expect that. it's pretty low priority so let's just leave it for now and maybe extend the TODO comment slightly just to explain that you tried and what was wrong with the attempt